### PR TITLE
ATO-1057 : Adds configured errors to ID token 

### PR DESCRIPTION
**Pull request by Ryan-Andrews99**
## What: 
- Our token endpoint can now return invalid responses
- These can be configured after the work done here: https://github.com/govuk-one-login/simulator/pull/105
## Why:
- This will allow RPs to test their error handling for the token endpoint
## Related PRs: 
- Branches off https://github.com/govuk-one-login/simulator/pull/105, so must be merged first 🙂

### CODE REVIEW

File: src/components/token/tests/helper/create-id-token.test.ts [hunk 1]
```diff
@@ -108,4 +112,110 @@ describe("createIdToken tests", () => {
...
+    const idToken = await createIdToken(mockAuthRequestParams, testAccessToken);
+
+    const ecTokenKey = await importPKCS8(EC_PRIVATE_TOKEN_SIGNING_KEY, "ES256");
+
+    await expect(jwtVerify(idToken, ecTokenKey)).rejects.toThrow(
+      errors.JWSSignatureVerificationFailed
+    );
+  });
+
+  it("returns a token with iat in the future if the client config has enabled TOKEN_NOT_VALID_YET enabled", async () => {
```
Comment (BenjaminWCO) [inline src/components/token/tests/helper/create-id-token.test.ts:148]:
```suggestion
  it("returns a token with iat in the future if the client config has enabled TOKEN_NOT_VALID_YET", async () => {
```

File: src/components/token/tests/helper/create-id-token.test.ts [hunk 2]
```diff
@@ -108,4 +112,110 @@ describe("createIdToken tests", () => {
...
+    tokenSigningAlgorithmSpy.mockReturnValue("ES256");
+    subSpy.mockReturnValue(testSubClaim);
+    clientIdSpy.mockReturnValue(testClientId);
+    idTokenErrorSpy.mockReturnValue(["TOKEN_NOT_VALID_YET"]);
+
+    const idToken = await createIdToken(mockAuthRequestParams, testAccessToken);
+
+    const payload = decodeTokenPart(idToken.split(".")[1]);
+    expect(payload.iat).toEqual(Math.floor(testTimestamp / 1000) + 86400);
```
Comment (BenjaminWCO) [inline src/components/token/tests/helper/create-id-token.test.ts:157]: might make it a bit clearer to re-name _testTimestamp_ to _testTimestampMilliseconds_ and create a constant called _secondsInADay_ for the offset?

File: tests/integration/token.test.ts [hunk 3]
```diff
@@ -432,52 +488,144 @@ describe("/token endpoint tests, invalid client assertion", () => {
...
+      .mockReturnValue(validAuthRequestParams);
+
+    const app = createApp();
+    const response = await request(app).post(TOKEN_ENDPOINT).send(validRequest);
+    const { id_token } = response.body;
+    const payload = decodeTokenPart(id_token.split(".")[1]);
+    expect(payload.iat).toBeGreaterThan(Date.now() / 1000);
+  });
+
+  it("returns an expired token if the client config has enabled TOKEN_EXPIRED enabled", async () => {
```
Comment (BenjaminWCO) [inline tests/integration/token.test.ts:574]: Same as above 

> "enabled X enabled"

File: src/components/token/tests/helper/create-id-token.test.ts (hunk 2 above)
Comment (Ryan-Andrews99) [reply src/components/token/tests/helper/create-id-token.test.ts]: Have renamed the timestamp to be `testTimestampMs` and created a constant `oneDayInSeconds` for the offset

File: tests/integration/token.test.ts (hunk 3 above)
Comment (Ryan-Andrews99) [reply tests/integration/token.test.ts]: Have fixed these aswell

File: src/components/token/tests/helper/create-id-token.test.ts (hunk 1 above)
Comment (BenjaminWCO) [reply src/components/token/tests/helper/create-id-token.test.ts:149]: token.test.ts looks fixed, but I think there's still a few in this file

Comment (Ryan-Andrews99) [reply src/components/token/tests/helper/create-id-token.test.ts]: Ahh yeah you're right, should be fixed now!

File: src/constants.ts [hunk 4]
```diff
@@ -68,3 +68,6 @@ export const ID_TOKEN_ERRORS = [
   "NONCE_NOT_MATCHING",
   "INCORRECT_VOT",
 ];
+
+export const INVALID_ISSUER = "https://example.com/identity-provider";
+export const oneDayInSeconds = 86400;
```
Comment (CarlyG55) [inline src/constants.ts:73]: nitpick but can we make this consistent casing?

File: src/components/token/helper/create-id-token.ts [hunk 5]
```diff
@@ -1,54 +1,83 @@
...
-  const iat = Math.floor(Date.now() / 1000);
-  const exp = iat + ID_TOKEN_EXPIRY;
+  const idTokenErrors = config.getIdTokenErrors();
+  const timeNow = Math.floor(Date.now() / 1000);
+  const iat = idTokenErrors.includes("TOKEN_NOT_VALID_YET")
+    ? timeNow + oneDayInSeconds
+    : timeNow;
+  const exp = idTokenErrors.includes("TOKEN_EXPIRED")
+    ? timeNow - oneDayInSeconds
+    : iat + ID_TOKEN_EXPIRY;
```
Comment (CarlyG55) [inline src/components/token/helper/create-id-token.ts:59]: could there be some conflict here if we have both errors? I think we want this to be `timeNow + ID_TOKEN_EXPIRY`?

**Comment by CarlyG55** [comment]
I'd also probably agree with the code smell here - nested ternaries are generally something to avoid

File: src/components/token/helper/create-id-token.ts (hunk 5 above)
Comment (Ryan-Andrews99) [reply src/components/token/helper/create-id-token.ts]: Good spot, have fixed!

File: src/constants.ts (hunk 4 above)
Comment (Ryan-Andrews99) [reply src/constants.ts]: Yep have fixed!

**Review by CarlyG55 — APPROVED** [review]

**Pull request merged**